llama_index/indices/managed/base.py

```diff
@@ -0,0 +1,75 @@
+"""Base Managed Service index.
+
+An index that that is built on top of a managed service.
+
+"""
+from abc import ABC, abstractmethod
+
+from typing import Any, Dict, List, Optional, Sequence, Tuple
+
+from llama_index.data_structs.data_structs import IndexDict
+from llama_index.indices.base import BaseIndex, IS
+from llama_index.indices.base_retriever import BaseRetriever
+from llama_index.indices.service_context import ServiceContext
+from llama_index.schema import BaseNode, Document
+from llama_index.storage.docstore.types import RefDocInfo
+
+
+class ManagedIndex(BaseIndex[IndexDict], ABC):
+    """Managed Index.
+    The managed service can index documents into a managed service.
+    How documents are structured into nodes is a detail for the managed service, and not exposed in this interface (although could be controlled by configuration parameters)
+
+    Args:
+        show_progress (bool): Whether to show tqdm progress bars. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        index_struct: Optional[IndexDict] = None,
+        show_progress: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        super().__init__(
+            nodes=None,
+            index_struct=index_struct,
+            service_context=None,
+            storage_context=None,
+            show_progress=show_progress,
+            **kwargs,
+        )
+
+    @abstractmethod
+    def insert(self, document: Document, **insert_kwargs: Any) -> None:
+        """Insert a document."""
+
+    @abstractmethod
+    def delete_ref_doc(
+        self, ref_doc_id: str, delete_from_docstore: bool = False, **delete_kwargs: Any
+    ) -> None:
+        """Delete a document and it's nodes by using ref_doc_id."""
+
+    @abstractmethod
+    def update_ref_doc(self, document: Document, **update_kwargs: Any) -> None:
+        """Update a document and it's corresponding nodes."""
+
+    @abstractmethod
+    def as_retriever(self, **kwargs: Any) -> BaseRetriever:
+        """Return a Retriever for this managed index."""
+
+    def _build_index_from_nodes(self, nodes: Sequence[BaseNode]) -> IndexDict:
+        """Build the index from nodes."""
+        raise NotImplementedError(
+            "_build_index_from_nodes not supported for a Managed index."
+        )
+
+    @abstractmethod
```

Comment:
Should probably remove this decorator, should help tests pass

Reply:
Done